lib/src/pubspec.dart

```diff
@@ -102,19 +100,64 @@ class Pubspec {
 
   String get issueTracker => _inner.issueTracker?.toString();
 
-  bool get hasDartSdkConstraint => _inner.environment.containsKey('sdk');
+  bool get hasDartSdkConstraint => sdkConstraintStatus.hasConstraint;
 
-  bool get shouldWarnDart2Constraint {
-    final vc = _inner.environment['sdk'];
-    if (vc == null) {
-      return false;
-    }
-    return !vc.allows(_dart2Last);
-  }
+  bool get shouldWarnDart2Constraint => !sdkConstraintStatus.enablesDart2Latest;
 
   bool get hasGitDependency =>
       _inner.dependencies.values.any((d) => d is GitDependency);
 
   bool get hasUnrestrictedGitDependency => _inner.dependencies.values
       .any((d) => d is GitDependency && (d.ref == null || d.ref.length < 40));
+
+  SdkConstraintStatus get sdkConstraintStatus =>
+      SdkConstraintStatus.fromSdkVersion(_inner.environment['sdk']);
+}
+
+final _range1 = VersionConstraint.parse('>=1.0.0 <2.0.0');
+final _range2 = VersionConstraint.parse('>=2.0.0 <3.0.0');
+final _range2Latest = VersionConstraint.parse('>=2.9999.0 <3.0.0');
+final _futureRange = VersionConstraint.parse('>=3.0.0');
+
+/// Detailed support coverage for the SDK constraint.
+class SdkConstraintStatus {
+  /// Whether it is non-empty, bounded constraint.
+  final bool hasConstraint;
+
+  /// Whether it allows anything from the ^1.0.0 range.
+  final bool enablesDart1;
```

Comment:
can't we make all of these private?

Reply:
that way I don't have to wonder if they are accessed from somewhere... and if we refactor it becomes a lot easier...

Also the names are not ideal... or well, they can be misunderstood... I'm not proposing better names :) hehe
I'm just proposing that we hide things other files have no business reading.